### PR TITLE
test: generate coverage profile during e2e tests

### LINKED ISSUE #900: end-to-end test generated coverage profile
### What is the version of your ORAS CLI

1.0.0

### What would you like to be added?

Run end-to-end tests based on [instrumented application](https://github.com/golang/proposal/blob/master/design/51430-revamp-code-coverage.md#running-instrumented-applications) and generate coverage report accordingly.

The released application will still not be instrumented.

### Why is this needed for ORAS?

Currently oras' [code coverage statistics](https://app.codecov.io/gh/oras-project/oras) are generated by unit testing, which is not accurate since:
1) By design, unit tests coverage report will only include files within the package. E.g if package `a` uses package `b` and `a` have unit test specs in `a_test.go`, unit test coverage report will not include package `b`. 
1) If a package is not included, Codecov report will skip the line counting, which means some **uncovered code may be silently ignored**. E.g. `oras.land/oras/cmd/oras/...` are not covered and not calculated in [oras 0.16](https://app.codecov.io/gh/oras-project/oras/tree/release-0.16). There are approximately 2K lines of golang code in oras but Codecov only show coverage for 550+ lines. 
1) Code coverage of end-to-end tests are not counted.

### Are you willing to submit PRs to contribute to this feature?

- [X] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR updates E2E test infra to generate test coverage profile when running e2e test. If the build is triggered by tagging and the tag starts with `v` (a release), e2e test will be ran directly on non-instrumented application.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #900.

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: test/e2e/scripts/e2e.sh [hunk 1]
```diff
@@ -43,25 +43,43 @@ fi
...
+fi
+
 echo " === run tests === "
-if ! ginkgo -r -p --succinct suite; then 
+ginkgo -r -p --succinct suite || fail=true
+
+if ! [ -z ${COVERAGE_DUMP_ROOT} ]; then
+  echo " === generating code cov report === "
+  go tool covdata textfmt -i="${e2e_root}/${COVERAGE_DUMP_ROOT}" -o ${e2e_root}/coverage.txt || true
+fi
```
Comment (TerryHowe) [inline test/e2e/scripts/e2e.sh:76]: Should there be a make target and .gitignore entry for this?

Comment (qweeah) [reply test/e2e/scripts/e2e.sh:77]: Changed as suggested.

**Review by TerryHowe — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**